### PR TITLE
Test Stokes Identities

### PR DESCRIPTION
**Pull request by alexfikl**
Add tests for some identities satisfied by the Stokeslet and the Stresslet on arbitrary geometries. They're given on page 2 of
https://services.math.duke.edu/~leili/teaching/duke/math660s18/lectures/lec25.pdf
or in Pozrikidis' tome of knowledge if available.

### CODE REVIEW

**Comment by alexfikl** [comment]
Convergence results in 2D on a 5-armed starfish for the Stokeslet

h                    | 5rd order             | EOC    | 3rd order              | EOC 
---------------------|-----------------------|--------|------------------------|-----
0.0625               | 7.347227267334338e-12 |        | 1.155847485638326e-09  |     
0.03125              | 6.995540192862465e-12 | 00.07  | 9.154592966619713e-10  | 0.33
0.015625             | 4.662484448865704e-15 | 10.55  | 3.639546560164975e-12  | 7.97
0.01041666           | 2.742219800926962e-17 | 12.66  | 1.136946078694970e-13  | 8.54
0.0078125            | 6.739665360935961e-19 | 12.88  | 7.701738440580504e-15  | 9.35

and the Stresslet

h                    | 5th order              | EOC   | 3rd order             | EOC      
---------------------|------------------------|-------|-----------------------|-----
0.0625               | 7.841613331065972e-08  |       | 5.169261460949249e-06 |                  
0.03125              | 7.335107515375984e-08  | 00.09 | 3.319709609159663e-06 | 0.63
0.015625             | 1.169692803219092e-10  | 09.29 | 4.870965283109148e-08 | 6.09 
0.01041666           | 1.323785820170431e-12  | 11.05 | 2.415487781430483e-09 | 7.40
0.0078125            | 4.911512430291654e-14  | 11.45 | 2.452037085419880e-10 | 7.95

No idea where the orders are coming from here. Looks like a quadrature `2n + 1` order (?), but I don't see why that error would dominate.

**Comment by alexfikl** [comment]
Just because it's so pretty, this is the error in the Stokeslet for `target_order = qbx_order = 3` and 4x oversampling and the finest refinement.

**Comment by inducer** [comment]
Thanks for doing this! So overall this seems to match results from #45 pretty well (IIUC)... something is causing accuracy to degrade severely in 3D for Stokes. In a way, I'm glad it's not a solve-only phenomenon. Building on the conjecture that the geometry derivatives are inaccurate: the QBX radius and center placement are derived from geometry derivatives, and if these are off, then I could very easily see things coming off the rails. Do things improve if you replace those with analytic/constant quantities?

**Comment by alexfikl** [comment]
> Building on the conjecture that the geometry derivatives are inaccurate: the QBX radius and center placement are derived from geometry derivatives, and if these are off, then I could very easily see things coming off the rails. Do things improve if you replace those with analytic/constant quantities?

Ah, completely forgot about those! I can easily see them actually having a bigger effect, as opposed to the area elements or other things in there. I'll try that out.

**Comment by alexfikl** [comment]
Ok, so time for another data dump with recent experiments. All of these are for the Stokeslet identity test with `target_order = qbx_order = 3` and 4x oversampling. Some of the numbers are slightly different than before because it's using `pytential.norm` to compute the relative errors.

First off, using analytical normals + a constant expansion radius (basically `h_max * Ones()`) for all elements seems to help a bit.

h              | Baseline       | EOC    | Normals        | EOC
:--------------|:---------------|:-------|:---------------|:------
1.28468859e+00 | 4.90769308e-04 |        | 2.60256347e-03 |
6.59786712e-01 | 1.23417237e-04 | 2.07   | 2.66794923e-04 | 3.42
3.30728970e-01 | 3.69521709e-05 | 1.75   | 3.71973115e-05 | 2.85
1.65394094e-01 | 1.54490224e-05 | 1.26   | 1.28638136e-05 | 1.53
8.26980033e-02 | 7.17999555e-06 | 1.11   | 5.59909245e-06 | 1.20

Next, I tried playing with the [fudge factor](https://github.com/inducer/pytential/blob/main/pytential/symbolic/primitives.py#L1011-L1016) for the expansion radii. By default it was set to `1` in 3D, so I tried bumping it up to `1.25` and `1.5`. This should still give h-convergence, right?

h              | 1.25           | EOC   | 1.5            | EOC    | 1.25 + Normals | EOC
:--------------|:---------------|:------|:---------------|:-------|:---------------|:-----
1.28468859e+00 | 1.42968485e-04 |       | 5.87162861e-05 |        | 2.25151878e-03 |
6.59786712e-01 | 2.47939804e-05 | 2.63  | 7.55798051e-06 | 3.08   | 1.95342839e-04 | 3.67
3.30728970e-01 | 5.37776584e-06 | 2.21  | 9.41696442e-07 | 3.02   | 1.23640652e-05 | 4.00
1.65394094e-01 | 1.96232164e-06 | 1.45  | 2.92240655e-07 | 1.69   | 1.79200673e-06 | 2.79
8.26980033e-02 | 8.54446355e-07 | 1.20  | 1.18752749e-07 | 1.30   | 6.53734352e-07 | 1.45

Also tried some of these combinations on the Stresslet identity with the same parameters. They seem to bottom out to first order a lot faster, but definitely better than the zero-order from https://github.com/inducer/pytential/pull/108#issuecomment-897917205.

h              | 1.25 + Normals | EOC
:--------------|:---------------|:------
1.28468859e+00 | 1.23881229e-02 |
6.59786712e-01 | 2.04313813e-03 | 2.70
3.30728970e-01 | 2.83060944e-04 | 2.86
1.65394094e-01 | 9.14596971e-05 | 1.63
8.26980033e-02 | 6.40306872e-05 | 0.51

**Comment by alexfikl** [comment]
@inducer This should be ready for a look now. It just adds two extra test cases for Stokes. Some takeaways:
* with the default `dim_fudge_factor = 0.5`, it requires 8x oversampling to look like it's converging
* bumping to `dim_fudge_factor = 1.0` (like in 2D), it converges with 4x oversampling (most of the results above are in this regime).

File: pytential/symbolic/primitives.py [hunk 1]
```diff
@@ -1056,6 +1056,17 @@ def h_max(ambient_dim, dim=None, dofdesc=None):
             cse_scope.DISCRETIZATION)
 
 
+def h_min(ambient_dim, dim=None, dofdesc=None):
+    """Defines a maximum element size in the discretization."""
```
Comment (inducer) [inline pytential/symbolic/primitives.py:1060]:
```suggestion
    """Yields an approximate minimum element size in the discretization."""
```

File: test/test_stokes.py [hunk 2]
```diff
@@ -44,6 +46,24 @@
 
 # {{{ test_exterior_stokes
 
+def rnorm2(actx, x, y, discr=None, p=None):
```
Comment (inducer) [inline test/test_stokes.py:49]: - Name arguments to designate one as reference?
- These look like two functions mashed into one with an `if`?

**Review by inducer — COMMENTED** [review]
Thanks! LGTM, just two minor nits, then it's good to go.

File: test/test_stokes.py (hunk 2 above)
Comment (alexfikl) [reply test/test_stokes.py]: > These look like two functions mashed into one with an if?

They were wearing a trenchcoat! How could you tell? :facepalm: 

Cleaned up in 45efe7b.

**Comment by inducer** [comment]
Thanks!

**Pull request merged**